<?xml version="1.0"?>
<robot name="cylinder" xmlns:xacro="http://ros.org/wiki/xacro">

 <!-- Define robot constants -->
 <xacro:property name="base_radius" value="0.12"/>
 <xacro:property name="base_length" value="0.093"/>
 
 <!-- Robot Base -->
 <link name="base_link">
  <inertial>
   <origin xyz="0 0 0" rpy="0 0 0"/>
   <mass value="1.0"/>
   <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
  </inertial>
  <visual>
   <geometry>
    <cylinder length="${base_length}" radius="${base_radius}"/>
   </geometry>
   <material name="blue">
    <color rgba="0.0 0.0 1.0 1.0"/>
   </material>
  </visual>
  <collision>
   <geometry>
    <cylinder length="${base_length}" radius="${base_radius}"/>
   </geometry>
  </collision>
 </link>

 <link name="base_footprint" />

 <joint name="base_joint" type="fixed">
  <parent link="base_link"/>
  <child link="base_footprint"/>
  <origin xyz="0 0 -0.052069" rpy="0 0 0"/>
 </joint>

 <joint name="lidar_joint" type="fixed">
  <parent link="base_link"/>
  <child link="base_scan"/>
  <origin xyz="0 0 0.0539" rpy="0 0 0"/>
 </joint>
 <link name="base_scan">
  <inertial>
   <origin xyz="0 0 0" rpy="0 0 0"/>
   <mass value="1.0"/>
   <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
  </inertial>
  <visual>
   <geometry>
    <cylinder length="0.0152" radius="0.017645"/>
   </geometry>
  </visual>
  <collision>
   <geometry>
    <cylinder length="0.0152" radius="0.017645"/>
   </geometry>
  </collision>
 </link>

 <joint name="imu_joint" type="fixed">
  <parent link="base_link"/>
  <child link="imu_link"/>
  <origin xyz="0 0 0.000569" rpy="0 0 0"/>
 </joint>
 <link name="imu_link">
  <inertial>
   <origin xyz="0 0 0" rpy="0 0 0"/>
   <mass value="1.0"/>
   <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
  </inertial>
  <visual>
   <geometry>
    <box size="0.01 0.01 0.01"/>
   </geometry>
  </visual>
  <collision>
   <geometry>
    <box size="0.01 0.01 0.01"/>
   </geometry>
  </collision>
 </link>
 <gazebo reference="imu_link">
  <sensor name="imu_sensor" type="imu">
   <plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
    <ros>
     <namespace>/demo</namespace>
     <remapping>~/out:=imu</remapping>
    </ros>
    <initial_orientation_as_reference>false</initial_orientation_as_reference>
   </plugin>
   <always_on>true</always_on>
   <update_rate>100</update_rate>
   <visualize>true</visualize>
   <imu>
    <angular_velocity>
     <x>
      <noise type="gaussian">
       <mean>0.0</mean>
       <stddev>2e-4</stddev>
       <bias_mean>0.000075</bias_mean>
       <bias_stddev>0.0000008</bias_stddev>
      </noise>
     </x>
     <y>
      <noise type="gaussian">
       <mean>0.0</mean>
       <stddev>2e-4</stddev>
       <bias_mean>0.0000075</bias_mean>
       <bias_stddev>0.000008</bias_stddev>
      </noise>
     </y>
     <z>
      <noise type="gaussian">
       <mean>0.0</mean>
       <stddev>2e-4</stddev>
       <bias_mean>0.0000075</bias_mean>
       <bias_stddev>0.000008</bias_stddev>
      </noise>
     </z>
    </angular_velocity>
    <linear_acceleration>
     <x>
      <noise type="gaussian">
       <mean>0.0</mean>
       <stddev>1.7e-2</stddev>
       <bias_mean>0.1</bias_mean>
       <bias_stddev>0.001</bias_stddev>
      </noise>
     </x>
     <y>
      <noise type="gaussian">
       <mean>0.0</mean>
       <stddev>1.7e-2</stddev>
       <bias_mean>0.1</bias_mean>
       <bias_stddev>0.001</bias_stddev>
      </noise>
     </y>
     <z>
      <noise type="gaussian">
       <mean>0.0</mean>
       <stddev>1.7e-2</stddev>
       <bias_mean>0.1</bias_mean>
       <bias_stddev>0.001</bias_stddev>
      </noise>
     </z>
    </linear_acceleration>
   </imu>
  </sensor>
 </gazebo>

 <joint name="wheel_left_joint" type="continuous">
  <origin xyz="0 0.09832 -0.016569" rpy="${radians(-90)} 0 0"/>
  <parent link="base_link"/>
  <child link="wheel_left"/>
  <axis xyz="0 0 1"/>
 </joint>

 <link name="wheel_left">
  <inertial>
   <origin xyz="0 0 0" rpy="0 0 0"/>
   <mass value="1.0"/>
   <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
  </inertial>
  <visual>
   <geometry>
    <cylinder radius="0.0355" length="0.0085"/>
   </geometry>
  </visual>
  <collision>
   <geometry>
    <cylinder radius="0.0355" length="0.0085"/>
   </geometry>
  </collision>
 </link>

 <joint name="wheel_right_joint" type="continuous">
  <origin xyz="0 -0.09832 -0.016569" rpy="${radians(-90)} 0 0"/>
  <parent link="base_link"/>
  <child link="wheel_right"/>
  <axis xyz="0 0 1"/>
 </joint>

 <link name="wheel_right">
  <inertial>
   <origin xyz="0 0 0" rpy="0 0 0"/>
   <mass value="1.0"/>
   <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
  </inertial>
  <visual>
   <geometry>
    <cylinder radius="0.0355" length="0.0085"/>
   </geometry>
  </visual>
  <collision>
   <geometry>
    <cylinder radius="0.0355" length="0.0085"/>
   </geometry>
  </collision>
 </link>

<gazebo>
  <plugin name='diff_drive' filename='libgazebo_ros_diff_drive.so'>
    <ros>
      <namespace>/demo</namespace>
    </ros>

    <!-- wheels -->
    <left_joint>wheel_left_joint</left_joint>
    <right_joint>wheel_right_joint</right_joint>

    <!-- kinematics -->
    <wheel_separation>0.19664</wheel_separation>
    <wheel_diameter>0.071</wheel_diameter>

    <!-- limits -->
    <max_wheel_torque>20</max_wheel_torque>
    <max_wheel_acceleration>1.0</max_wheel_acceleration>

    <!-- output -->
    <publish_odom>true</publish_odom>
    <publish_odom_tf>false</publish_odom_tf>
    <publish_wheel_tf>true</publish_wheel_tf>

    <odometry_frame>odom</odometry_frame>
    <robot_base_frame>base_link</robot_base_frame>
  </plugin>
</gazebo>
 
 <joint name="front_ball_joint" type="fixed">
  <origin xyz="0.10 0 -0.042069" rpy="0 0 0"/>
  <parent link="base_link"/>
  <child link="front_ball_link"/>
 </joint>
 
 <link name="front_ball_link">
  <visual>
   <origin xyz="0 0 0" rpy="0 0 0"/>
   <geometry>
    <sphere radius="0.01"/>
   </geometry>
  </visual>
  <collision>
   <origin xyz="0 0 0" rpy="0 0 0"/>
   <geometry>
    <sphere radius="0.01"/>
   </geometry>
  </collision>
 </link>
  
 <joint name="back_ball_joint" type="fixed">
  <origin xyz="-0.10 0 -0.042069" rpy="0 0 0"/>
  <parent link="base_link"/>
  <child link="back_ball_link"/>
 </joint>
 
 <link name="back_ball_link">
  <visual>
   <origin xyz="0 0 0" rpy="0 0 0"/>
   <geometry>
    <sphere radius="0.01"/>
   </geometry>
  </visual>
  <collision>
   <origin xyz="0 0 0" rpy="0 0 0"/>
   <geometry>
    <sphere radius="0.01"/>
   </geometry>
  </collision>
 </link>
  
 <joint name="camera_joint" type="fixed">
  <parent link="base_link"/>
  <child link="camera"/>
  <origin xyz="0.12 0 0.025" rpy="${radians(-90)} ${radians(0)} ${radians(-90)}"/>
 </joint>

 <link name="camera">
  <visual>
   <geometry>
    <box size="0.01 0.01 0.01"/>
   </geometry>
  </visual>
  <collision>
   <geometry>
    <box size="0.01 0.01 0.01"/>
   </geometry>
  </collision>
 </link>

</robot>
